### PR TITLE
chore: add new error type OpaqueWithText

### PR DESCRIPTION
**Pull request by ajewellamz**
*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithydafny/DafnyApiCodegen.java [hunk 1]
```diff
@@ -1879,6 +1879,8 @@ public TokenTree generateModeledErrorDataType() {
           "// The Opaque error, used for native, extern, wrapped or unknown errors"
         ),
         Token.of("| Opaque(obj: object)"),
+        Token.of("// A better Opaque, with a visible string representation."),
+        Token.of("| OpaqueWithText(obj: object, objMessage : string)"),
         // Helper error for use with `extern`
         Token.of("type OpaqueError = e: Error | e.Opaque? witness *")
```
Comment (texastony) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithydafny/DafnyApiCodegen.java]: I think we can classify `OpaqueError` as either Opaque or OpaqueWithText.
```suggestion
        Token.of("type OpaqueError = e: Error | e.Opaque? || e.OpaqueWithText? witness *")
```

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToDafnyAwsV2.java [hunk 2]
```diff
@@ -705,7 +705,7 @@ MethodSpec generateConvertOpaqueServiceError() {
         "nativeValue"
       )
       .addStatement(
-        "return $T.create_Opaque(nativeValue)",
+        "return $T.create_OpaqueWithText(nativeValue, dafny.DafnySequence.asString(nativeValue.getMessage()))",
```
Comment (texastony) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToDafnyAwsV2.java:708]: Blocking: Regrettably, we are going to need both an option for OpaqueWithText and Opaque.

Though maybe not in ToDafnyAws,
it probably simplifies our complexity to just know it is ALWAYS an option.

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToNativeAwsV2.java [hunk 3]
```diff
@@ -29,7 +29,7 @@
 import software.amazon.polymorph.smithyjava.nameresolver.AwsSdkDafnyV2;
 import software.amazon.polymorph.smithyjava.nameresolver.AwsSdkNativeV2;
 import software.amazon.polymorph.smithyjava.nameresolver.Dafny;
-import software.amazon.polymorph.smithyjava.unmodeled.OpaqueError;
+import software.amazon.polymorph.smithyjava.unmodeled.OpaqueWithTextError;
```
Comment (texastony) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToNativeAwsV2.java]: Blocking: We will always need both... at least until we Major Version.

**Review by texastony — DISMISSED** [review]
It looks Rust and Java have forked.
Rust does, what I think, is the correct behavior, and creates new methods to handle the new type.

Java does the wrong thing,
and replaces the old methods with methods to only handle the new type.

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToDafnyAwsV2.java (hunk 2 above)
Comment (ajewellamz) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToDafnyAwsV2.java:708]: I don't understand. How can the return type be optional? Where do we put the configuration?

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToNativeAwsV2.java (hunk 3 above)
Comment (ajewellamz) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithyjava/generator/awssdk/v2/ToNativeAwsV2.java]: I'm confused. How does adding an import change anything?

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java [hunk 4]
```diff
@@ -206,6 +206,69 @@ def __eq__(self, other: Any) -> bool:
                         getattr(self, a) == getattr(other, a)
                         for a in attributes
                     )
+
+            class OpaqueWithTextError(ApiError[Literal["OpaqueWithTextError"]]):
+                code: Literal["OpaqueWithTextError"] = "OpaqueWithTextError"
+                obj: Any  # As an OpaqueWithTextError, type of obj is unknown
+
```
Comment (lucasmcdonald3) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java]:
```suggestion
                obj: Any  # As an OpaqueWithTextError, type of obj is unknown
                obj_message: str # obj_message is a message representing the details of obj
```

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java [hunk 5]
```diff
@@ -206,6 +206,69 @@ def __eq__(self, other: Any) -> bool:
...
+                def __init__(
+                    self,
+                    *,
+                    obj,
+                    objMessage
```
Comment (lucasmcdonald3) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java]:
```suggestion
                    obj_message
```

Comment (lucasmcdonald3) [reply codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java]: Ditto for other native types in this class; let's prefer snake-case for Smithy-generated code

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java [hunk 6]
```diff
@@ -206,6 +206,69 @@ def __eq__(self, other: Any) -> bool:
                         getattr(self, a) == getattr(other, a)
                         for a in attributes
                     )
+
+            class OpaqueWithTextError(ApiError[Literal["OpaqueWithTextError"]]):
```
Comment (lucasmcdonald3) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java:210]: I think you will need to update error handling to add native/Dafny conversions for this. It's an unmodelled error, so conversions are not generated automatically.
Check out these 2 spots where this is done for other unmodelled errors:

* https://github.com/smithy-lang/smithy-dafny/blob/main-1.x/codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java#L408

* https://github.com/smithy-lang/smithy-dafny/blob/d8870f5151475523fcc5a675e2128eb0e4ff5d03/codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/DafnyPythonLocalServiceProtocolGenerator.java#L459

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java [hunk 7]
```diff
@@ -415,6 +479,16 @@ private void generateSmithyErrorToDafnyErrorBlock(
         codegenContext
       )
     );
+    writer.write(
```
Comment (lucasmcdonald3) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java:482]: This does native -> Dafny,  but you need to add the Dafny -> native conversion here:

https://github.com/smithy-lang/smithy-dafny/blob/d8870f5151475523fcc5a675e2128eb0e4ff5d03/codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/DafnyPythonLocalServiceProtocolGenerator.java#L459

File: codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java [hunk 8]
```diff
@@ -206,6 +206,70 @@ def __eq__(self, other: Any) -> bool:
...
+                    return {
+                        'message': self.message,
+                        'code': self.code,
+                        'obj': self.obj,
+                        'obj_message': self.obj_message,
+                    }
+
+                @staticmethod
+                def from_dict(d: Dict[str, Any]) -> "OpaqueError":
+                    ""\"Creates a OpaqueError from a dictionary.
```
Comment (lucasmcdonald3) [inline codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithypython/localservice/customize/ErrorsFileWriter.java]:
```suggestion
                @staticmethod
                def from_dict(d: Dict[str, Any]) -> "OpaqueWithTextError":
                    ""\"Creates a OpaqueWithTextError from a dictionary.
```

**Review by texastony — APPROVED** [review]
I have high confidence that this achieves it's objective based on testing I pushed here:
https://github.com/aws/aws-cryptographic-material-providers-library/actions/runs/11749257486/job/32735108567?pr=969

I have high confidence this does not break anyone,
based on local testing I did against the DB-ESDK-Java at 3.7.0 and a local build of the the MPL created with this.

~~OH... my testing was invalid... hang on...~~ 

I fixed my local testing and I know that we are good!

**Pull request merged**